verticapy/learn/causal_model.py

```diff
@@ -0,0 +1,343 @@
+		if self.pdag != None:
+			return CausalModel.visualize(self.pdag, directed=True)
+
+	# ---#
+	@staticmethod
+	def visualize(adj_dict, directed=True):
+		nodes = list(adj_dict)
+		directed_edges = set([(x, y) for x in nodes for y in adj_dict[x]])
+		try:
+			import networkx as nx
+			import matplotlib.pyplot as plt
+			from pyvis.network import Network
+
+			if directed:
+				G = nx.DiGraph()
+			else:
+				G = nx.Graph()
+			G.add_nodes_from(nodes)
+			G.add_edges_from(directed_edges)
+
+			try:
+				shell = get_ipython().__class__.__name__
+				if shell == 'ZMQInteractiveShell':
+					notebook = True
+				else:
+					notebook = False
+			except NameError:
+				notebook = False
+			
+			nt = Network('500px', '500px', notebook=notebook, directed = directed)
+			nt.from_nx(G)
+			nt.save_graph('nx.html')
+			
+			from IPython.display import IFrame
+			return IFrame(src='./nx.html', width=700, height=700)
+			
+		except ModuleNotFoundError:
+			warnings.warn("Could not produce graph because one or more of the following modules are missing: [networkx, matplotlib, pyvis]")
+					
+
+
+class PC(CausalModel):
+	# ---#
+	def __init__(self, input_relation):
+		super().__init__(input_relation)
+
+	# ---#
+	def build_skeleton(self, method='stable', max_cond_vars=5, significance_level = 0.01, show_progress=True):
+		"""
+	----------------------------------------------------------------------------------------
+	Generate the skeleton for the causal network.
+	
+	Parameters
+	----------
+	method: str, optional
+		stable: Produces the same result despite the ordering of variables.
+		orig: Follows the original pc algorithm.
+
+	max_cond_vars: int, optional
+		non-zero value that determines how many conditional variables are used, representing
+		the expected number of parents any node has. The higher the number the longer the test
+		will take to run, but the more precise the result will be.
+
+	significance_level: float, optional
+		p-value that determines independence.
+
+	show_progress: bool, optional
+		The value to control if the progress bar is displayed.
+
+	Returns
+	--------
+	skeleton_edges: dict
+		undirected edges in an adjacency dictionary
+	sepsets: dict
+		maps frozenset edge pairs to there corresponding seperation set
+		"""
+
+		columns = [col[1:-1] for col in self.vdf._VERTICAPY_VARIABLES_['columns']]
+		all_pairs = set(itertools.combinations(columns, 2))
+		max_cond_vars = min(len(columns), max_cond_vars+1) 
+		sepsets = {} 
+		skeleton_edges = {x: set(columns) - set([x]) for x in columns}
+		
+		cit = conditional_chi_square # Conditional Independence Test
+		show_progress = show_progress and vp.options["tqdm"] # Verticapy Override option
+		if show_progress:
+			progress_bar = tqdm(total=max_cond_vars)
+			progress_bar.set_description("Using n conditional variables: 0")
+
+		# Discover the skeleton
+		for cond_set_size in range(max_cond_vars):
+			if method == 'stable':
+				neighbors = {x: skeleton_edges[x].copy() for x in columns}
```

Comment:
Except this line, every other line is the same for `method == 'stable'` and `method == 'orig'`; therefor you can make an if-statement only for this line and consolidate the rest. Btw, you can make sure `method` has a valid value before the for-loop.

Reply:
Ah, I see that there is another difference between the usage of `neighbors` and `skeleton_edges`. So you can copy the right one to a temp variable to consolidate the code, but it would be worth it if they are always small. I'll leave the judgment to you. That said, I think it would still be cleaner if you validate `method` before the for-loop.

Reply:
@zacandcheese  please correct it in your code